### DISCUSSION TITLE
blog/aspnet-core-api-with-jwt-authentication/

### DISCUSSION THREAD

**Discussion by giscus** [codewithmukesh/discussions#39] [General]
# blog/aspnet-core-api-with-jwt-authentication/

In this article, we will talk in detail about various API versioning techniques and How to implement API Versioning in ASP.NET Core 3.1 Applications.

https://codewithmukesh.com/blog/aspnet-core-api-with-jwt-authentication/

**Comment by jaynaras** [comment]
Do the following if you would like to use Net 8.0.
Step 1.
net8.0
Install following packages.
 Microsoft.AspNetCore.Authentication.JwtBearer Version="8.0.0"
 Microsoft.AspNetCore.Identity.EntityFrameworkCore Version="8.0.0"
 Microsoft.EntityFrameworkCore Version="8.0.0"
 Microsoft.EntityFrameworkCore.Design Version="8.0.0"
 Microsoft.EntityFrameworkCore.SqlServer Version="8.0.0"
 <PackageReferense Include="Microsoft.EntityFrameworkCore.Tools Version="8.0.0"
 Microsoft.VisualStudio.Web.CodeGeneration.Design Version="8.0.0"
 Swashbuckle.AspNetCore Version="6.6.2"
 System.IdentityModel.Tokens.Jwt Version="8.0.0"
Step 2.
-- add the following two lines at Constants/Authorization.cs
   public const string default_firstname = "firstname";
   public const string default_lastname = "lastname";
Step 3.
-- modify this line to add these properties on Contexts/ApplicationDbContextSeed.cs
var defaultUser = new ApplicationUser
{ UserName = Authorization.default_username, FirstName = Authorization.default_firstname, LastName = Authorization.default_lastname, Email = Authorization.default_email, EmailConfirmed = true, PhoneNumberConfirmed = true };
Step 4.
-- modify the appsettings.json
-- add XYZABC at the end of value of key property
"JWT": {
  "key": "C1CF4B7DC4C4175B6618DE4F55CA4XYZABC",
  "Issuer": "SecureApi",
  "Audience": "SecureApiUser",
  "DurationInMinutes": 60
},

Then it should work as expected.

**Reply by jaynaras** [reply]
FYI:
-- One more change in Program.cs. (since there is not startup.cs on .net 8.0)
-- added startup.cs info on to program.cs as follows:
var builder = WebApplication.CreateBuilder(args);

// Add services to the container.

//Configuration from AppSettings
builder.Services.Configure<JWT>(builder.Configuration.GetSection("JWT"));

//User Manager Service
builder.Services.AddIdentity<ApplicationUser, IdentityRole>().AddEntityFrameworkStores<ApplicationDbContext>();
builder.Services.AddScoped<IUserService, UserService>();
//Adding DB Context with MSSQL
builder.Services.AddDbContext<ApplicationDbContext>(options =>
    options.UseSqlServer(
        builder.Configuration.GetConnectionString("DefaultConnection"),
        b => b.MigrationsAssembly(typeof(ApplicationDbContext).Assembly.FullName)));
//Adding Athentication - JWT
builder.Services.AddAuthentication(options =>
{
    options.DefaultAuthenticateScheme = JwtBearerDefaults.AuthenticationScheme;
    options.DefaultChallengeScheme = JwtBearerDefaults.AuthenticationScheme;
})

    .AddJwtBearer(o =>
    {
        o.RequireHttpsMetadata = false;
        o.SaveToken = false;
        o.TokenValidationParameters = new TokenValidationParameters
        {
            ValidateIssuerSigningKey = true,
            ValidateIssuer = true,
            ValidateAudience = true,
            ValidateLifetime = true,
            ClockSkew = TimeSpan.Zero,

            ValidIssuer = builder.Configuration["JWT:Issuer"],
            ValidAudience = builder.Configuration["JWT:Audience"],
            IssuerSigningKey = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(builder.Configuration["JWT:Key"]))
        };
    });

builder.Services.AddControllers();
// Learn more about configuring Swagger/OpenAPI at https://aka.ms/aspnetcore/swashbuckle
builder.Services.AddEndpointsApiExplorer();
builder.Services.AddSwaggerGen();

var app = builder.Build();

using (var scope = app.Services.CreateScope())
{
    var services = scope.ServiceProvider;
    var loggerFactory = services.GetRequiredService<ILoggerFactory>();
    try
    {
        //Seed Default Users
        var userManager = services.GetRequiredService<UserManager<ApplicationUser>>();
        var roleManager = services.GetRequiredService<RoleManager<IdentityRole>>();
        await ApplicationDbContextSeed.SeedEssentialsAsync(userManager, roleManager);
    }
    catch (Exception ex)
    {
        var logger = loggerFactory.CreateLogger<Program>();
        logger.LogError(ex, "An error occurred seeding the DB.");
    }
}

// Configure the HTTP request pipeline.
if (app.Environment.IsDevelopment())
{
    app.UseSwagger();
    app.UseSwaggerUI();
}

app.UseHttpsRedirection();
app.UseRouting();
app.UseAuthentication();
app.UseAuthorization();

app.MapControllers();

app.Run();
